kolibri/core/discovery/utils/network/client.py

```diff
@@ -13,84 +18,213 @@
     "subset_of_users_device": False,
 }
 
+DEFAULT_CONNECT_TIMEOUT = 5
+DEFAULT_READ_TIMEOUT = 60
+# default read timeout when within a job
+DEFAULT_ASYNC_READ_TIMEOUT = 30
+# when the network client tries variations of a url, that means the overall length of time it takes
+# is multiplied by the number of variations, so for synchronous operations (in a HTTP request) we
+# make the overall timeout ~= the DEFAULT_READ_TIMEOUT
+DEFAULT_SYNC_READ_TIMEOUT = DEFAULT_READ_TIMEOUT / (len(HTTP_PORTS) + len(HTTPS_PORTS))
 
-class NetworkClient(object):
-    DEFAULT_TIMEOUT_IN_SECS = 5
 
-    def __init__(self, base_url=None, address=None, timeout=None, **kwargs):
-        """If an explicit base_url is already known, provide that. If a vague address is provided, we can try to infer the base_url"""
-        if not base_url and not address:
-            raise Exception(
-                "You must provide either a `base_url` or `address` argument"
+class NetworkClient(requests.Session):
+    __slots__ = ("base_url", "timeout", "session", "device_info", "remote_ip")
+
+    def __init__(self, base_url, timeout=None):
+        """
+        If an explicit base_url is already known, provide that. If only a vague address is known,
+        `build_from_address` can build a client to determine the actual `base_url`
+        :param base_url: The fully composed URL for a network location, without path
+        :param timeout: A timeout value in seconds or tuple for (connect, read)
+        :type timeout: float|tuple
+        """
+        super(NetworkClient, self).__init__()
+
+        self.base_url = base_url
+        self.timeout = timeout or (DEFAULT_CONNECT_TIMEOUT, DEFAULT_READ_TIMEOUT)
+        self.session = None
+        self.device_info = None
+        self.remote_ip = None
+
+    @classmethod
+    def build_for_address(cls, address, timeout=None):
```

Comment:
Originally, this logic always occurred in the constructor of the `NetworkClient`. By moving it to a factory method, it makes the client a little more useful as a utility